### PR TITLE
Add support for float:toFixedString() and float:toExpString()

### LINKED ISSUE #35181: Langlib functions to provide control over conversion of numbers to strings
Please refer to : https://github.com/ballerina-platform/ballerina-spec/issues/770

### LINKED ISSUE #35541: Implement `toFixedString` langlib function for floats
Please refer to : https://github.com/ballerina-platform/ballerina-spec/commit/c673dbba066c88a9651a634c11113c70b0b1b888

Spec Issue: https://github.com/ballerina-platform/ballerina-spec/issues/770

### LINKED ISSUE #35542: Implement `toExpString` langlib function for floats
Please efer to this commit: https://github.com/ballerina-platform/ballerina-spec/commit/c673dbba066c88a9651a634c11113c70b0b1b888

Spec Issue: https://github.com/ballerina-platform/ballerina-spec/issues/770

### PR DESCRIPTION
**Pull request by lasinicl**
## Purpose
$title

Fixes #35541
Fixes #35542
Fixes #35181

## Check List 
- [x] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [x] Added necessary tests
  - [x] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+import static io.ballerina.runtime.internal.util.exceptions.BallerinaErrorReasons.getModulePrefixedReason;
+
+/**
+ * Native implementation of lang.float:toExpString(float, int?).
+ *
+ * @since 2.0.0
+ */
+public class ToExpString {
+    public static BString toExpString(double x, Object fractionDigits) {
+        // If `x` is NaN or infinite, the result will be the same as `value:toString`.
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:42]: Can we move the common validation in toExpString() and toFixedString() to a util ?

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+            double multipliedValue = xAbsValue;
+
+            while (multipliedValue < 1) {
+                multipliedValue = xAbsValue * Math.pow(10, tens + 1);
+                tens++;
+            }
+        }
+
+        DecimalFormat decimalFormat = new DecimalFormat();
+        decimalFormat.setRoundingMode(RoundingMode.HALF_EVEN);
```
Comment (KRVPerera) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:83]: We can create object later. It is being replaced anyway.

File: langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal [hunk 3]
```diff
@@ -415,6 +415,720 @@ function testFromStringNegative() {
...
+
+    float c = -5.7;
+    string d = float:toFixedString(c, 9223372036854775807);
+    assertEquality("-5.7", d);
+    assertEquality("-5.7", c.toFixedString(int:MAX_VALUE));
+}
+
+function testToFixedStringWithMorePositiveFloats() {
+    assertEquality("45362.1233399999982794", float:toFixedString(45362.12334, 16));
+    assertEquality("0.1237469219000000", float:toFixedString(0.1237469219, 16));
```
Comment (KRVPerera) [inline langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:640]: Shall we add one or more test case to reflect rounding mode in `float:toFixedString`

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+        }
+
+        if (noOfFractionDigits > Integer.MAX_VALUE) {
+            return StringUtils.fromString(StringUtils.getStringValue(x, null));
+        }
+
+        int fracDig = (int) noOfFractionDigits;
+
+        BigDecimal numberBigDecimal = new BigDecimal(x);
+        numberBigDecimal  = numberBigDecimal.setScale(fracDig, RoundingMode.HALF_EVEN);
```
Comment (lochana-chathura) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java:63]:
```suggestion
        int fracDigits = (int) noOfFractionDigits;

        BigDecimal numberBigDecimal = new BigDecimal(x);
        numberBigDecimal  = numberBigDecimal.setScale(fracDigits, RoundingMode.HALF_EVEN);
```

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 5]
```diff
@@ -0,0 +1,114 @@
...
+import java.math.RoundingMode;
+import java.text.DecimalFormat;
+
+import static io.ballerina.runtime.api.constants.RuntimeConstants.FLOAT_LANG_LIB;
+import static io.ballerina.runtime.internal.util.exceptions.BallerinaErrorReasons.getModulePrefixedReason;
+
+/**
+ * Native implementation of lang.float:toExpString(float, int?).
+ *
+ * @since 2.0.0
```
Comment (gimantha) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:38]:
```suggestion
 * @since 2201.1.0
```

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+import static io.ballerina.runtime.api.constants.RuntimeConstants.FLOAT_LANG_LIB;
+import static io.ballerina.runtime.internal.util.exceptions.BallerinaErrorReasons.getModulePrefixedReason;
+
+/**
+ * Native implementation of lang.float:toFixedString(float, int?).
+ *
+ * @since 2.0.0
```
Comment (gimantha) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java:37]:
```suggestion
 * @since 2201.1.0
```

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        int indexOfExp = res.lastIndexOf("E");
+        String firstSection = res.substring(0, indexOfExp);
+        int idxOfDecimalPoint = firstSection.lastIndexOf(".");
+        if (idxOfDecimalPoint == firstSection.length() - 1) {
+            firstSection = res.substring(0, idxOfDecimalPoint);
+        }
+        String secondSection = res.substring(indexOfExp + 1);
+        int p = Integer.parseInt(secondSection);
+        if (p >= 0) {
+            secondSection = "e+" + secondSection;
```
Comment (KRVPerera) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:107]: IMO this is wrong simple `e` means natural exponent right...

Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:107]: The spec says `e`.
https://github.com/ballerina-platform/ballerina-spec/blob/master/lang/lib/float.bal#L298

File: langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal [hunk 8]
```diff
@@ -415,6 +415,733 @@ function testFromStringNegative() {
...
+    a = -0x1.e412904862198p4;
+    assertEquality("-3.02545321300000012e+1", a.toExpString(17));
+    assertEquality("-3.0254532130e+1", a.toExpString(10));
+    assertEquality("-3.0254e+1", a.toExpString(4));
+
+    b = -0x1.cbe3p17;
+    assertEquality("-2.35462000000000000e+5", b.toExpString(17));
+    assertEquality("-2.3546200000e+5", b.toExpString(10));
+    assertEquality("-2.3546e+5", b.toExpString(4));
+}
```
Comment (lochana-chathura) [inline langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:1143]: Spec says, 
> If `x` is zero, the exponent is zero. A zero exponent is represented with a `+` sign.

Did we cover that? 
I'm getting a runtime exception for that.
```bal
    float a = 0;
    string s = a.toExpString(());
```

```
Error Output: error: java.lang.NumberFormatException {"message":"Character N is neither a decimal digit number, decimal point, nor "e" notation exponential mark."}
	at ballerina.lang.float.0:toExpString(float.bal:428)
```

Comment (lasinicl) [reply langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:1143]: Did cover, but seems not properly handled when the fractionDigits is (). will fix.

Comment (lasinicl) [reply langlib/langlib-test/src/test/resources/test-src/floatlib_test.bal:1151]: Fixed in d86a63e

File: langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibFloatTest.java [hunk 9]
```diff
@@ -123,5 +123,55 @@ public void testFromString(String function) {
...
-    }  
+    }
+
+    @Test(dataProvider = "functionsToTestToFixedString")
+    public void testToFixedString(String function) {
+        BRunUtil.invoke(compileResult, function);
+    }
+
+    @DataProvider
+    public  Object[] functionsToTestToFixedString() {
```
Comment (KRVPerera) [inline langlib/langlib-test/src/test/java/org/ballerinalang/langlib/test/LangLibFloatTest.java:134]: For `toFixedString` this output has a exponent
```Ballerina
float number = 12321321312321;
io:println(number.toFixedString(11431241242));
```

```
1.2321321312321E13
```

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 10]
```diff
@@ -0,0 +1,120 @@
...
+        } else {
+            secondSection = "e" + secondSection;
+        }
+
+        return StringUtils.fromString(firstSection + secondSection);
```
Comment (gimantha) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:118]: Can we not generate the pattern with `#` and simply generate the ExpString? May be we can address it in a separate PR

**Review by KRVPerera — DISMISSED** [review]
LGTM

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 10 above)
Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:118]: Will address this separately if the PR does not need any further changes.

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    public static BString toFixedString(double x, Object fractionDigits) {
+        // If `x` is NaN or infinite, the result will be the same as `value:toString`.
+        // If fractionalDigits is `()`, use the minimum number of digits required to accurately represent the value.
+        BString res = FloatUtils.getBStringValue(x, fractionDigits == null);
+        if (res != null) {
+            return StringUtils.fromString(StringUtils.getStringValue(x, null));
+        }
+
+        long noOfFractionDigits = (long) fractionDigits;
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java:54]: Can `fractionDigits` ever be null?

Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java:54]: Yes. That's handled.

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 12]
```diff
@@ -0,0 +1,120 @@
...
+        }
+
+        // Handle very large int values since they might cause overflows.
+        if (noOfFractionDigits > Integer.MAX_VALUE ||
+                (noOfFractionDigits * Long.bitCount(noOfFractionDigits) > Integer.MAX_VALUE)) {
+            // Maximum value a double can represent is around 1.7*10^308.
+            noOfFractionDigits = 308;
+        }
+
+        int exponent = (int) noOfFractionDigits;
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:79]: Any reason to bring `noOfFractionDigits` as long and then cast to int?

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 13]
```diff
@@ -0,0 +1,120 @@
...
+            while (multipliedValue < 1) {
+                multipliedValue = xAbsValue * Math.pow(10, tens + 1);
+                tens++;
+            }
+        }
+
+        BigDecimal numberBigDecimal = new BigDecimal(x);
+        if (xAbsValue != 0 && xAbsValue < 1) {
+            numberBigDecimal  = numberBigDecimal.setScale(exponent + tens, RoundingMode.HALF_EVEN);
+
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:94]:
```suggestion
```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/FloatUtils.java [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+
+import io.ballerina.runtime.api.utils.StringUtils;
+import io.ballerina.runtime.api.values.BString;
+
+/**
+ * Common utility methods used for float operations.
+ *
+ * @since 2201.1.0
+ */
+public class FloatUtils {
```
Comment (lochana-chathura) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/FloatUtils.java:29]:
```suggestion
public class FloatUtils {

```

File: bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/FloatUtils.java [hunk 15]
```diff
@@ -0,0 +1,36 @@
...
+    public static BString getBStringValue(double x) {
```
Comment (lochana-chathura) [inline bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/FloatUtils.java:30]: For me `getBStringValue` is too generic. HBT `getBStringIfInfiniteOrNaN`?

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+            return StringUtils.fromString(StringUtils.getStringValue(x, null));
+        }
+
+        long noOfFractionDigits;
+
+        // If fractionalDigits is `()`, use the minimum number of digits required to accurately represent the value.
+        double xAbsValue = Math.abs(x);
+        if (fractionDigits == null) {
+            if (xAbsValue == 0) {
+                noOfFractionDigits = BigDecimal.valueOf(xAbsValue).scale();
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:56]: Wouldn't `BigDecimal.valueOf(xAbsValue).scale()` for 0 will always be 1

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+
+        // Panic if noOfFractionDigits < 0.
+        if (noOfFractionDigits < 0) {
+            throw ErrorCreator.createError(getModulePrefixedReason(FLOAT_LANG_LIB,
+                    BallerinaErrorReasons.INVALID_FRACTION_DIGITS_ERROR),
+                    BLangExceptionHelper.getErrorDetails(RuntimeErrors.INVALID_FRACTION_DIGITS));
+        }
+
+        // Handle very large int values since they might cause overflows.
+        if (noOfFractionDigits > Integer.MAX_VALUE ||
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:73]: Same validation for both functions.

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java [hunk 18]
```diff
@@ -0,0 +1,75 @@
...
+
+        if (fractionDigits == null) {
+            return StringUtils.fromString(BigDecimal.valueOf(x).toPlainString());
+        }
+
+        long noOfFractionDigits = (long) fractionDigits;
+
+        // Panic if noOfFractionDigits < 0.
+        if (noOfFractionDigits < 0) {
+            throw ErrorCreator.createError(getModulePrefixedReason(FLOAT_LANG_LIB,
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToFixedString.java:58]: Can move this to util and share the error in both cases

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 10 above)
Comment (pcnfernando) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:118]: Refer the java.text.DecimalFormat when you do

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 16 above)
Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:56]: yes. Since the value is a `float` that's okay right?

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 17 above)
Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:73]: This was done intentionally since the `fractionDigits` is a ballerina `int` (which is Java `long`) and scaling the value to a given number of `fractionDigits` using BigDecimal require a java `int`

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 12 above)
Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:79]: The reason is mentioned in https://github.com/ballerina-platform/ballerina-lang/pull/35703#discussion_r846054633

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 16 above)
Comment (pcnfernando) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:56]: My point is; we can directly assign 1 instead of the function call

Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:50]: ah yes! will change.

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 19]
```diff
@@ -0,0 +1,111 @@
...
+ *
+ * @since 2201.1.0
+ */
+public class ToExpString {
+
+    public static BString toExpString(double x, Object fractionDigits) {
+        // If `x` is NaN or infinite, the result will be the same as `value:toString`.
+        BString str = FloatUtils.getBStringIfInfiniteOrNaN(x);
+        if (str != null) {
+            return StringUtils.fromString(StringUtils.getStringValue(x, null));
```
Comment (lochana-chathura) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:41]: sorry. just saw this.
```suggestion
            return str;
```

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 17 above)
Comment (pcnfernando) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:65]: No, my concern was this `if` check is the same for both functions. Hence check can be moved to utils, isn't it?

Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:65]: ah sorry! will update.

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 10 above)
Comment (gimantha) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:108]: https://github.com/ballerina-platform/ballerina-lang/issues/35757

**Review by KRVPerera — APPROVED** [review]

**Review by lochana-chathura — APPROVED** [review]

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java [hunk 20]
```diff
@@ -0,0 +1,110 @@
...
+                int integerPart = (int) (Math.log10(xAbsValue));
+                noOfFractionDigits = BigDecimal.valueOf(xAbsValue / Math.pow(10, integerPart)).scale();
+            }
+        } else {
+            noOfFractionDigits = (long) fractionDigits;
+        }
+
+        // Panic if noOfFractionDigits < 0.
+        if (noOfFractionDigits < 0) {
+            throw ErrorUtils.createInvalidFractionDigitsError();
```
Comment (pcnfernando) [inline langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:61]: Shall we add tests for error throwing

**Review by pcnfernando — APPROVED** [review]
LGTM

File: langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java (hunk 20 above)
Comment (gimantha) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:61]: Lets address them here: https://github.com/ballerina-platform/ballerina-lang/issues/35757

Comment (lasinicl) [reply langlib/lang.float/src/main/java/org/ballerinalang/langlib/floatingpoint/ToExpString.java:61]: i have already added tests
